### DISCUSSION TITLE
.NET 8 + Blazor In OC Module?

### DISCUSSION THREAD

**Discussion by websitewill** [OrchardCMS/OrchardCore#15593] [Q&A]
Is there an updated example somewhere that shows how to add a new Blazor component to an existing OC Module?
My project has a project for the main system, and various module projects. I want to rewrite part of one module (which was some incredibly funky javascript work) in Blazor. Basically just one view in that module (which has a few views) has something of a client-side experience that was written in knockout (old). I want to move that to Blazor as a PoC - and pave the way for further adoption if possible.

I am having trouble getting everything to line up - seems the last time I touched Startup.cs has been a few years and the later OC examples are a bit different. I just updated to .net 8 yesterday and have the main app + theme + modules all working. Now, just can't seem to get Blazor to plug in nicely.

I keep getting "The type or namespace name 'Components' does not exist in the namespace 'redacted' (are you missing an assembly reference?).  I'm not. :) Oddly, when I open the error site and hover, VS clearly knows that the object is and I can right-click and go to definition. So something in vs build is complaining for nothing?

To get here, I added a new Blazor Web App (.net 8) project. Looks to be setup like the default project, with my own simple Components/Message.razor file. No Pages - I just want something simple to start.

_Imports.razor has @using for the component.

Orchard module (server side) startup.cs has this.
`
 endpointRouteBuilder.MapRazorComponents<Message>()
     .AddInteractiveServerRenderMode()
     .AddInteractiveWebAssemblyRenderMode()
     .AddAdditionalAssemblies(typeof(redacted.Blazor._Imports).Assembly);

In the module ConfigureServices method I have this.
`
    services.AddRazorComponents()
        .AddInteractiveServerComponents()
        .AddInteractiveWebAssemblyComponents();
`
I have tried various combinations of this stuff and found nothing that gets rid of the error and lets it work.

Module has a route to
Test/TestBlazor and I have a views folder
Views/Test/TestBlazor.cshtml

That file looks like this
`
@using redacted.Blazor.Components
﻿@T["Hello from Module 1!"]

<component type="typeof(Message)" render-mode="ServerPrerendered" />
`

Any ideas or updated PoCs to peek at?

**Comment by Piedone** [comment]
While I won't be able to help with Blazor, be sure to check these out:

- https://github.com/OrchardCMS/OrchardCore/pull/15573
- https://github.com/OrchardCMS/OrchardCore/pull/15213
- https://github.com/OrchardCMS/OrchardCore.Samples/pull/45

**Comment by websitewill** [comment]
Thanks @Piedone ,

Your first link seems to hint that what I am doing is entirely possible (and I don't know why it wouldn't be).

I had actually cloned those other sample repos over the last few days and none of them seem to be showing what I am wanting. The "Add Blazor guide" seems to address the decoupled CMS - so it's only using the Targets package and everything else is basically a standalone app (blazor).

What I need is to integrate Blazor into an existing module using the Full CMS.

**Reply by websitewill** [reply]
I am making a little progress. Everything is compiling (had to completely uninstall VS and reinstall to get it to even acknowledge the namespaces). Also, had to pay attention to how the project was setup (specifically the Build Action seems to be Content).

Now, it all compiles and my portal is running again, but the razor component still won't render. In my logs I am seeing this;
System.InvalidOperationException: The current thread is not associated with the Dispatcher. Use InvokeAsync() to switch execution to the Dispatcher when triggering rendering or component state.

Can confirm that if I simply remove 

@(await Html.RenderComponentAsync<Message>(RenderMode.ServerPrerendered)) 

from the Test.cshtml page in the module, then the module works. 
Same happens if I try to use this:

`<component type="typeof(Message)" render-mode="ServerPrerendered" />`

So, something to do with the blazor component itself..

**Reply by websitewill** [reply]
Still struggling here.
Things compile, if I don't attempt to load the blazor component the module's content loads just fine.
But as soon as I add:
`<component type="typeof(Message)" render-mode="WebAssemblyPrerendered" />`
or
@(await Html.RenderComponentAsync<Message>(RenderMode.ServerPrerendered))
or... - any of the other ways I've found online to try, I end up with this error.

I spun up a vanilla asp.net MVC site. Added a Razor Web App project (deleted the server, kept the client) and was able to add the Counter.razor to the Index.cshtml and get it rendered with ease. Is this just some nuance with Orchard Core that I am missing?

Screenshot of the sample for proof:
![image](https://github.com/OrchardCMS/OrchardCore/assets/19608705/799838bd-d77e-4fec-88b0-8a4c9ac99416)

**Reply by websitewill** [reply]
Further update.
I tried to add the Blazor component directly to the theme (before my @RenderSection for content.
Just added 
```
<h3>BLAZOR TEST</h3>
<component type="typeof(Message)" render-mode="WebAssemblyPrerendered" />
<h3>BLAZOR TEST</h3>
```

And it worked fine. Exact same Blazor but used int he Layout versus the Module works. Is this due to how the modules are rendered from the layout? Investigating - but at least a little progress (and proof that the module itself works in OC in some form).

**Comment by websitewill** [comment]
Starting to wonder if there is just some issue in the guts of OCs implementation of RazorPage<TModel>?
I have that plain MVC app and I am able to add my razor component directly in the _Layout, or in Index.cshtml.
It works - doesn't in Orchard. 
The difference? RazorPage<TModel>?
Not sure of the root cause yet but I see that OC has completely changed RenderSection and RenderSectionAsync. OC has those declared as "new" so the base MVC code doesn't fire.

Clearly OC is doing something very different here than what regular MVC does - I'm guessing to facilitate modules and themes for OC, but somewhere in the guts of that, it's fundamentally different, which is preventing Blazor components from working.

Any ideas?

**Comment by websitewill** [comment]
Should I just create an issue from this discussion? Was holding off until I figured out if it was me or just something in OC. Starting to think this is an OC issue. @Piedone, thoughts?

**Comment by websitewill** [comment]
Looks like is an open issue that addresses a related concern. Pulled the 1.9.0 preview packages and the Blazor component now renders. Guess just a matter of time. Closing.

**Reply by Piedone** [reply]
Great, then!
